topology/probes/blockdev/blockdev.go

```diff
@@ -0,0 +1,462 @@
+		}
+		if blockdev.Type == multipathType {
+			groupNode = p.addGroup(blockdev, parentWWN)
+		}
+	} else {
+		groupNode = p.addGroup(blockdev, parentWWN)
+	}
+
+	p.Lock()
+	defer p.Unlock()
+	var graphMetaData graph.Metadata
+
+	if _, ok := p.blockdevMap[blockdev.Name]; ok {
+		return groupNode
+	}
+
+	p.Ctx.Graph.Lock()
+	defer p.Ctx.Graph.Unlock()
+
+	node := p.Ctx.Graph.LookupFirstNode(graph.Metadata{"Path": p.getPath(blockdev)})
+
+	if node == nil {
+		graphMetaData = p.getMetaData(blockdev, childCount, parentWWN)
+		var err error
+		if node, err = p.Ctx.Graph.NewNode(graph.GenID(), graphMetaData); err != nil {
+			p.Ctx.Logger.Error(err)
+			return nil
+		}
+
+		// If there is a WWN - check to see if there are other paths to the same block device.
+		if blockdev.WWN != "" {
+			multiPathNodes := p.Ctx.Graph.GetNodes(graph.Metadata{"BlockID": p.getID(blockdev)})
+
+			for _, multiPathNode := range multiPathNodes {
+				topology.AddLink(p.Ctx.Graph, multiPathNode, node, "multipath", nil)
+			}
+		}
+
+		topology.AddOwnershipLink(p.Ctx.Graph, groupNode, node, nil)
+
+		// Link physical disks and DVD/rom devices to the blockdevGroupName
+		if blockdev.Type == "disk" || blockdev.Type == "rom" {
+			topology.AddLink(p.Ctx.Graph, p.Groups[blockdevGroupName], node, "connected", nil)
+		}
+
+	}
+
+	for i := range blockdev.Children {
+		childNode := p.Ctx.Graph.LookupFirstNode(graph.Metadata{"Name": p.getName(blockdev.Children[i])})
+		if childNode != nil {
+			topology.AddLink(p.Ctx.Graph, childNode, node, "connected", nil)
+		}
+	}
+
+	p.blockdevMap[blockdev.Name] = blockdevInfo{
+		ID:   blockdev.Name,
+		Node: node,
+	}
+	return groupNode
+}
+
+func (p *ProbeHandler) unregisterBlockdev(id string) {
+	p.Lock()
+	defer p.Unlock()
+
+	info, ok := p.blockdevMap[id]
+	if !ok {
+		return
+	}
+
+	p.Ctx.Graph.Lock()
+	if err := p.Ctx.Graph.DelNode(info.Node); err != nil {
+		p.Ctx.Graph.Unlock()
+		p.Ctx.Logger.Error(err)
+		return
+	}
+	p.Ctx.Graph.Unlock()
+
+	delete(p.blockdevMap, id)
+}
+
+func (p *ProbeHandler) connect() error {
+	var (
+		cmdOut []byte
+		err    error
+		result Devices
+	)
+
+	testFile := os.Getenv("SKYDIVE_BLOCKDEV_TEST_FILE")
+
+	if testFile != "" {
+		if cmdOut, err = exec.Command("cat", testFile).Output(); err != nil {
+			p.Ctx.Logger.Error(err)
+			os.Exit(1)
+		}
+	} else {
+		if cmdOut, err = exec.Command("lsblk", "-pO", "--json").Output(); err != nil {
+			p.Ctx.Logger.Error(err)
+			os.Exit(1)
+		}
+	}
+
+	jsonBlockdevs := string(cmdOut[:])
+	if err = json.Unmarshal([]byte(jsonBlockdevs), &result); err != nil {
+		p.Ctx.Logger.Error(err)
+		os.Exit(1)
+	}
+
+	// loop through the devices in the current map to make sure they haven't
+	// been removed
+	for _, current := range p.blockdevMap {
+		p.deleteIfRemoved(current, result.Blockdevices)
+	}
+
+	for i := range result.Blockdevices {
+		p.registerBlockdev(result.Blockdevices[i], "")
+	}
+
+	return nil
+}
+
+// Start the probe
+func (p *ProbeHandler) Start() error {
```

Comment:
You could make NewProbe (renamed from Init) return a probe wrapper as in https://github.com/skydive-project/skydive/blob/master/topology/probes/docker/docker.go#L281
Therefore, you don't need to implement Start and Stop, but only the Do method. That would give something like :

```
 func (p *ProbeHandler) Do(ctx context.Context, wg *sync.WaitGroup) error {
    p.addGroupByName(blockdevGroupName, "")
    ...

    wg.Add(1)
    go func() {
        defer wg.Done()

        after := time.After(30*time.Second)
        for {
            if err := p.connect(); err != nil {
                p.Ctx.Logger.Error(err)
                return
            }

            select {
                case <- ctx.Done():
                    return
                case <-after:
            }
        }
    }
```

Reply:
Ok - thanks.